EMS/core-bundle/src/Service/DataService.php

```diff
@@ -1348,43 +1348,49 @@ public function putBack(ContentType $contentType, string $ouuid): ?int
+    public function updateDataStructure(FieldType $meta, DataField $dataField, int $parentKey = 0): void
     {
-        // no need to generate the structure for subfields
-        $isContainer = true;
+        if (null === $fieldType = $dataField->getFieldType()) {
+            return;
+        }
 
-        if (null !== $dataField->getFieldType()) {
-            $dataFieldType = $this->formRegistry->getType($dataField->getFieldType()->getType())->getInnerType();
+        $dataFieldType = $this->formRegistry->getType($fieldType->getType())->getInnerType();
+        if (!$dataFieldType instanceof DataFieldType && !self::isInternalField($fieldType->getName())) {
+            $this->logger->warning('service.data.not_a_data_field', ['field_name' => $fieldType->getName()]);
 
-            if ($dataFieldType instanceof DataFieldType) {
-                $isContainer = $dataFieldType->isContainer();
-            } elseif (!DataService::isInternalField($dataField->getFieldType()->getName())) {
-                $this->logger->warning('service.data.not_a_data_field', [
-                    'field_name' => $dataField->getFieldType()->getName(),
-                ]);
-            }
+            return;
         }
 
-        if ($isContainer) {
-            /** @var FieldType $field */
-            foreach ($meta->getChildren() as $key => $field) {
-                // no need to generate the structure for delete field
-                if (!$field->getDeleted()) {
-                    $child = $dataField->__get('ems_'.$field->getName());
-                    if (null == $child) {
-                        $child = new DataField();
-                        $child->setFieldType($field);
-                        $child->setOrderKey($field->getOrderKey());
-                        $child->setParent($dataField);
-                        $dataField->addChild($child, $key);
-                        if (isset($field->getDisplayOptions()['defaultValue'])) {
-                            $child->setEncodedText($field->getDisplayOptions()['defaultValue']);
-                        }
-                    }
-                    if (0 != \strcmp($field->getType(), CollectionFieldType::class)) {
-                        $this->updateDataStructure($field, $child);
-                    }
-                }
+        if ($dataFieldType instanceof DataFieldType && !$dataFieldType::isContainer()) {
+            return;
+        }
+
+        foreach ($meta->getChildren() as $key => $field) {
+            if ($field->getDeleted() || null !== $dataField->__get('ems_'.$field->getName())) {
+                continue;
+            }
+
+            if (FormFieldType::class === $field->getType()) {
+                /** @var FormFieldType $formFieldType */
+                $formFieldType = $this->formRegistry->getType($field->getType())->getInnerType();
+                $formMeta = $formFieldType->getReferredFieldType($field);
```

Comment:
this is the fixing, getting the formEntity and build all children.

Passing the current key as $parentKey argument.